src/databricks/labs/lsql/structs.py

```diff
@@ -0,0 +1,166 @@
+        fields = ",".join(f.as_sql() for f in self.fields)
+        return f"STRUCT<{fields}>"
+
+    def as_schema(self) -> str:
+        fields = []
+        for field in self.fields:
+            not_null = "" if field.nullable else " NOT NULL"
+            fields.append(f"{field.name} {field.type.as_sql()}{not_null}")
+        return ", ".join(fields)
+
+
+class StructInference:
+    _PRIMITIVES: ClassVar[dict[type, str]] = {
+        str: "STRING",
+        int: "LONG",
+        bool: "BOOLEAN",
+        float: "FLOAT",
+        datetime.date: "DATE",
+        datetime.datetime: "TIMESTAMP",
+    }
+
+    def as_ddl(self, type_ref: type) -> str:
+        v = self._infer(type_ref, [])
+        return v.as_sql()
+
+    def as_schema(self, type_ref: type) -> str:
+        v = self._infer(type_ref, [])
+        if hasattr(v, "as_schema"):
+            return v.as_schema()
+        raise StructInferError(f"Cannot generate schema for {type_ref}")
+
+    def _infer(self, type_ref: type, path: list[str]) -> SqlType:
+        if dataclasses.is_dataclass(type_ref):
+            return self._infer_struct(type_ref, path)
+        if isinstance(type_ref, enum.EnumMeta):
+            return self._infer_primitive(str, path)
+        if type_ref in self._PRIMITIVES:
+            return self._infer_primitive(type_ref, path)
+        if type_ref is list:
+            raise StructInferError("Cannot determine element type of list. Rewrite as: list[XXX]")
+        if type_ref is set:
+            raise StructInferError("Cannot determine element type of set. Rewrite as: set[XXX]")
+        if type_ref is dict:
+            raise StructInferError("Cannot determine key and value types of dict. Rewrite as: dict[XXX, YYY]")
+        return self._infer_generic(type_ref, path)
+
+    def _infer_primitive(self, type_ref: type, path: list[str]) -> PrimitiveType:
+        if type_ref in self._PRIMITIVES:
+            return PrimitiveType(self._PRIMITIVES[type_ref])
+        raise StructInferError(f'{".".join(path)}: unknown: {type_ref}')
+
+    def _infer_generic(self, type_ref: type, path: list[str]) -> SqlType:
+        # pylint: disable-next=import-outside-toplevel
+        from typing import (  # type: ignore[attr-defined]
+            _GenericAlias,
+            _UnionGenericAlias,
+        )
+
+        if isinstance(type_ref, (types.UnionType, _UnionGenericAlias)):  # type: ignore[attr-defined]
+            return self._infer_nullable(type_ref, path)
+        if isinstance(type_ref, (_GenericAlias, types.GenericAlias)):  # type: ignore[attr-defined]
+            if type_ref.__origin__ in (dict, list) or isinstance(type_ref, types.GenericAlias):
+                return self._infer_container(type_ref, path)
+        prefix = ".".join(path)
+        if prefix:
+            prefix = f"{prefix}: "
+        raise StructInferError(f"{prefix}unsupported type: {type_ref.__name__}")
+
+    def _infer_nullable(self, type_ref: type, path: list[str]) -> SqlType:
```

Comment:
I think we infer a union type here, but we only support the nullable as outcome:

```suggestion
    def _infer_union(self, type_ref: type, path: list[str]) -> SqlType:
```

Reply:
we're using spark types here